library/modules/Units.cpp

```diff
@@ -802,7 +800,15 @@ df::unit *Units::getUnit (const int32_t index)
     return vector_get(world->units.all, index);
 }
 
-// returns index of creature actually read or -1 if no creature can be found
+df::unit *Units::getUnitFromHFID(int32_t hf_id)
+{
+    auto hf = df::historical_figure::find(hf_id);
+    if (hf)
+        return df::unit::find(hf->unit_id);
+    else
+        return vector_get(world->units.all, linear_index(world->units.all, &df::unit::hist_figure_id, hf_id));
```

Comment:
in which situations will the else branch produce a useful result?

Reply:
If the unit is missing a hf entry, the unit can still be found by its hf id. IDK why the entry wouldn't exist, but the Bay12 code accounts for it.